common/testing/testlogger/testlogger.go

```diff
@@ -0,0 +1,514 @@
+	tl.state.mu.expectations[e.lvl].Remove(e.e)
+}
+
+func (tl *TestLogger) shouldFailTest(level Level, msg string, tags []tag.Tag) bool {
+	expectations := tl.state.mu.expectations[level]
+	for e := expectations.Front(); e != nil; e = e.Next() {
+		m, ok := e.Value.(matcher)
+		if !ok {
+			tl.state.t.Fatalf("Bug in TestLogger: invalid %T value in matcher list", e.Value)
+		}
+		if m.Matches(msg, tags) {
+			return tl.state.mode == FailOnExpectedErrorOnly
+		}
+	}
+	return tl.state.mode == FailOnAnyUnexpectedError
+}
+
+// PanicOnError overrides the behavior of this logger. It returns the previous value
+// so that it can be restored later.
+func (tl *TestLogger) PanicOnError(v bool) bool {
+	return tl.state.panicOnError.Swap(v)
+}
+
+// DPanic implements log.Logger.
+func (tl *TestLogger) DPanic(msg string, tags ...tag.Tag) {
+	tl.state.mu.RLock()
+	defer tl.state.mu.RUnlock()
+	if tl.state.mu.closed {
+		return
+	}
+	if tl.tags != nil {
+		tags = append(tags, tl.tags...)
+	}
+	// note, actual panic'ing in wrapped is turned off so we can control.
+	tl.wrapped.DPanic(msg, tags...)
+	if tl.state.panicOnDPanic.Load() && tl.shouldFailTest(Error, msg, tags) {
+		tl.state.t.Helper()
+		panic(failureMessage("DPanic", msg, tags))
+	}
+}
+
+// Debug implements log.Logger.
+func (tl *TestLogger) Debug(msg string, tags ...tag.Tag) {
+	tl.state.mu.RLock()
+	defer tl.state.mu.RUnlock()
+	if tl.state.mu.closed {
+		return
+	}
+	if tl.tags != nil {
+		tags = append(tags, tl.tags...)
+	}
+	_ = tl.shouldFailTest(Debug, msg, tags)
+	tl.wrapped.Debug(msg, append(tags, tl.tags...)...)
+}
+
+// Error implements log.Logger.
+func (tl *TestLogger) Error(msg string, tags ...tag.Tag) {
+	tl.state.mu.RLock()
+	if tl.state.mu.closed {
+		tl.state.mu.RUnlock()
+		return
+	}
+	if tl.tags != nil {
+		tags = append(tags, tl.tags...)
+	}
+	if !tl.shouldFailTest(Error, msg, tags) {
+		tl.wrapped.Error(msg, tags...)
+		tl.state.mu.RUnlock()
+		return
+	}
+	tl.state.mu.RUnlock()
+
+	if tl.state.panicOnError.Load() {
+		tl.state.t.Helper()
+		tl.wrapped.Error(msg, tags...)
+		panic(failureMessage("Error", msg, tags))
+	}
+
+	// Labeling the error as unexpected; so it can easily be identified later.
+	tl.wrapped.Error(errorMessage("Error", msg), tags...)
+
+	// Not panic'ing, so marking the test as failed.
+	tl.state.mu.Lock()
+	tl.state.mu.failed = true
+	tl.state.mu.Unlock()
+}
+
+// Fatal implements log.Logger.
+func (tl *TestLogger) Fatal(msg string, tags ...tag.Tag) {
+	tl.state.mu.RLock()
+	defer tl.state.mu.RUnlock()
+	if tl.state.mu.closed {
+		return
+	}
+	if tl.tags != nil {
+		tags = append(tags, tl.tags...)
+	}
+	tl.state.t.Helper()
+	tl.state.t.Fatal(failureMessage("Fatal", msg, tags))
+}
+
+// Info implements log.Logger.
+func (tl *TestLogger) Info(msg string, tags ...tag.Tag) {
+	tl.state.mu.RLock()
+	defer tl.state.mu.RUnlock()
+	if tl.state.mu.closed {
+		return
+	}
+	if tl.tags != nil {
+		tags = append(tags, tl.tags...)
+	}
+	_ = tl.shouldFailTest(Info, msg, tags)
```

Comment:
Calls to this in levels below ERROR are vestigial and can be removed. Their purpose has long since been excised from the original codebase

Reply:
👍 Alright, I can remove them. I thought the purpose for them was logging.